File: graphs/kruskal/kruskal.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

class UnionFind {
public:
    vector<int> root;
    vector<int> rank;

    UnionFind(int size) : root(size), rank(size, 1) {
        for (int i = 0; i < size; ++i) {
            root[i] = i;
        }
    }

    int find(int x) {
        if (root[x] != x) {
            root[x] = find(root[x]);
        }
        return root[x];
    }

    void unionNodes(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX != rootY) {
            if (rank[rootX] > rank[rootY]) {
                root[rootY] = rootX;
            } else if (rank[rootX] < rank[rootY]) {
                root[rootX] = rootY;
            } else {
                root[rootY] = rootX;
                rank[rootX] += 1;
            }
        }
    }
};

vector<vector<int>> kruskal(int n, map<int, vector<pair<int,int>>>& graph) {
    // Collect all edges
    vector<vector<int>> edges;
    for (auto item = graph.begin(); item != graph.end(); item++) {
        int a = item->first;
        vector<pair<int,int>>& neighbors = item->second;
        for (auto& neighbor:neighbors){
            int b = neighbor.first;
            int weight = neighbor.second;
            edges.push_back({weight, a, b});
        }
    }

    // Sort edges by weight
    sort(edges.begin(), edges.end());

    vector<vector<int>> mst;
    UnionFind uf(n);

    // Iterate through sorted edges
    for (auto item = edges.begin(); item != edges.end(); item++){
        int weight = (*item)[0];
        int a = (*item)[1];
        int b = (*item)[2];
        // If the nodes are not already connected
        if (uf.find(a) != uf.find(b)){
            // Connect them
            uf.unionNodes(a, b);
            mst.push_back({a, b, weight});
        }
    }

    return mst;
}

int main() {
    map<int, vector<pair<int, int>>> graph1;
    graph1[0] = {{1, 4}, {7, 8}};
    graph1[1] = {{2, 8}, {7, 11}};
    graph1[2] = {{5, 4}, {3, 7}};
    graph1[3] = {{4, 9}, {5, 14}};
    graph1[4] = {};
    graph1[5] = {{4, 10}};
    graph1[6] = {{5, 2}};
    graph1[7] = {{6, 1}, {8, 7}};
    graph1[8] = {{2, 2}, {6, 6}};
    auto mst = kruskal(9, graph1);

    cout << "Minimum spanning tree edges:" << endl;
    for (auto mst_edge:mst){
        cout << mst_edge[0] << " " << mst_edge[1] << " " << mst_edge[2] << endl;
    }

    return 0;
}

// Expected: 
// 7 6 1
// 6 5 2
// 8 2 2
// 0 1 4
// 2 5 4
// 2 3 7
// 0 7 8
// 3 4 9

File: graphs/kruskal/kruskal.py
from collections import defaultdict

class UnionFind:
    def __init__(self, size):
        self.root = [i for i in range(size)]
        self.rank = [1 for i in range(size)]
    
    def find(self, x):
        if self.root[x] != x:
            self.root[x] = self.find(self.root[x])
        return self.root[x]
    
    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.root[rootY] = rootX
            elif self.rank[rootX] < self.rank[rootY]:
                self.root[rootX] = rootY
            else:
                self.root[rootY] = rootX
                self.rank[rootX] += 1

def kruskal(n:int, graph:defaultdict(list)):
    # collect all edges
    edges = []
    for a, neighbors in graph.items():
        for b, weight in neighbors:
            edges.append((weight, a, b))
    
    # sort edges by weight
    edges.sort()

    mst = []
    uf = UnionFind(size=n)

    # iterate through sorted edges
    for weight, a, b in edges:
        # if the nodes are not already connected
        if uf.find(a) != uf.find(b):
            # connected them
            uf.union(a, b)
            mst.append((a, b, weight))
    
    return mst

if __name__ == "__main__":
    # Test case 1
    graph1 = defaultdict(list)
    graph1[0] = [(1, 1), (2, 2)]
    graph1[1] = [(0, 1), (2, 3)]
    graph1[2] = [(0, 2), (1, 3)]
    assert kruskal(3, graph1) == [(0, 1, 1), (0, 2, 2)]

    # Test case 2
    graph2 = defaultdict(list)
    graph2[0] = [(1, 4), (2, 3)]
    graph2[1] = [(0, 4), (3, 2)]
    graph2[2] = [(0, 3), (3, 5)]
    graph2[3] = [(1, 2), (2, 5)]
    assert kruskal(4, graph2) == [(1, 3, 2), (0, 2, 3), (0, 1, 4)]
    
    # Test case 3
    graph3 = defaultdict(list)
    graph3[0] = [(1, 4), (7, 8)]
    graph3[1] = [(2, 8), (7, 11)]
    graph3[2] = [(5, 4), (3, 7)]
    graph3[3] = [(4, 9), (5, 14)]
    graph3[4] = []
    graph3[5] = [(4, 10)]
    graph3[6] = [(5, 2)]
    graph3[7] = [(6, 1), (8, 7)]
    graph3[8] = [(2, 2), (6, 6)]
    mst = kruskal(9, graph3)
    assert kruskal(9, graph3) == [(7, 6, 1), (6, 5, 2), (8, 2, 2), (0, 1, 4), (2, 5, 4), (2, 3, 7), (0, 7, 8), (3, 4, 9)]
    
    print("All test cases passed!")